### ISSUE TITLE
Support post migration for already-known posts and media

### ISSUE THREAD

**Issue by nemobis** [mastodon/mastodon#19902]
### Pitch

When migrating an account to a new instance, the user should have the possibility to also migrate their own posts with their media. Because this presents all sorts of issues, such a feature should be opt-in for target instances and minimise the resources and trust required (if it proves possible at all).

In this feature request I suggest that, as first step in this feature, the target instance could allow the user to "import" the posts which are **already cached** in said instance. In other words, no *content* would need to move anywhere.

As a step 0, posts could be somehow displayed alongside the new profile instead of redirected to a known-deprecated or known-dead instance. For example, https://chaos.social/web/@Senficon@respublicae.eu/109254331876095530 currently redirects to https://respublicae.eu/@Senficon/109254331848944112 (when logged in), even though https://respublicae.eu/@Senficon redirects to https://chaos.social/@senficon . (I've not tested this with the new unified UI in v4.0.) The instance could just stop redirecting, and instead display the post from its own cache.

Step 1 would be some sort of database schema change which would allow to update cached posts as being now first-class citizens of the instance, perhaps with some sort of alias ID which the instance would create to claim them as its own.

Step 2 would be any change on third instances. These could be notified that a user has moved, and that they can optionally update their references to that user so that new requests related to its posts (fetching them or interacting with them) are forwarded to the new instance/new alias IDs (if created).

### Motivation

Because instances come and go, some users care about keeping their posts online after migrating: https://github.com/mastodon/mastodon/issues/12423 . There are arguably three goals/components in this: 1) keeping the content available for new users; 2) preventing link rot for web visitors; 3) keeping old interactions on third-party instances working. They don't necessarily need to be solved all at once.

Importing old posts is currently possible with some server-side imports or [scripts](https://pleromabot.robertoszek.xyz/pleroma-bot/) which just repost old posts as new. This solves (1) but now (2) and (3).

Focusing on posts which are already cached in the target instance would allow to address separately the issue of backfilling/fetching/migrating old posts, with the attendant issues of trust and verification (how to verify that the posts aren't made up, perhaps by someone who has maliciously taken over the domain? etc.).

**Comment by trwnh** [comment]
Depends on #17030 and #10745 -- currently posts are tied to usernames and domains and a lot of fundamental work would need to be done to decouple post identifiers from that.

**Comment by thiagomgd** [comment]
To also add on why that's important:
Not only because instances come and go - so right now I don't really wanna bet on any smaller instance, since I might lose all my posts.

But also because the messaging around people trying mastodon is always "you can start on any instance to try it out, and then move on later". Right now this is not true, at least according to users' expectations. If you tell me I can try any instance and move to another one later, I would really want to take my posts with me.

**Comment by DRKSLV** [comment]
Remote Posts are already in the same database table as regular posts.
So that's kinda handled.

To do whats proposed, an account just needs to verify *that* it was the target of a move, and *when* it was the target of a move.
The moving server should deliver the MOVE activity to the moved-to server too, which records the time in account_aliases.moved_on, which should be NULL before a move.
Then, on a users profile, it shows all posts `WHERE ( account_id = [local-user] ) OR ( account_id=[remote-user] AND updated_at <  [move-time] )` (just a sketch)

This would, in my eyes, be a trivial change (but i have no idea what moving parts might be involved so lmao)

And this approach is easily extendable: The only thing missing for a full migration is to iterate through the old profile and force-pull an arbritrary selection of posts. (which can be implemented any number of ways, or not at all, and be customizable for instance owners)

Another concern that will eventually arise is the situation of multiple accounts being moved into one (which is allowed now)
Which means displaying a profile page might have to do something more cursed like LEFT JOINing the alias table into statuses then checking WHERE acc_id=[local_user] OR (moved_on IS NOT NULL AND updated_at<moved_on)

But more tricky, the situation of an account being moved twice:
It would be difficult to verify that the account was previously moved between two remote servers!
(which i think would only be solvable through cryptographically signing the move activity?? (jwt?) 
and publishing it somehow or sending it to the next server during the second move????
but since account keys can change.... ehhhh)

In any case. The simple version of this issue would be a pretty good step forward i think.
If the old server is still up, you could even move it to a second account, then, when you leave that account too,
move the first AND the second account to the third account.
Not available if the first server is down, but whatever. 
You can't do anything on behalf of the first account when the first one is down, anyway.

I don't know if there is anything substantial in the way of this, i am not familiar with the codebase, haven't really dealt with activity pub etc..

**Comment by DRKSLV** [comment]
The reason i like this proposal: (and actually wrote all of this into an issue of my own)
If a user is moving between instances, it is likely that they are already known on the target instance, people are following them, and many of the posts are *already* in the database.

My first post was really detailed, i am sorry but i am sick of recieving increasingly more unrealistic emails from the Post Migration #12423 thread. Lmao
Specific proposals like this seem to be the way to go (easy to think about and close/act on)

**Comment by DRKSLV** [comment]
> Depends on #17030 and #10745 -- currently posts are tied to usernames and domains and a lot of fundamental work would need to be done to decouple post identifiers from that.

Yeah, the local user id would have to be in the account_aliases table too...
To effectively merge into the statuses table etc.
Is this a problem currently?

**Comment by Danie10** [comment]
The replicate option may be good, because I just want to mention that posts already made are linked to by others, often externally in blogs etc. So be deleting original posts (not much can be done for an instance closing down) we could have masses of broken links on the Internet.
